File: Segundo/PDOO/Tema 1/Civitas/src/main/java/civitas/Tablero.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;

/**
 *
 * @author jmramirez
 */
public class Tablero {
    private int numCasillaCarcel;
    private ArrayList<Casilla> casillas;
    private int porSalida;
    private boolean tieneJuez;
    
    public Tablero(int carcel) {        //Constructor al que se le pasa por parámetros la casilla carcel
        if (carcel >= 1) {      //if para comprobar la validez de la casilla carcel
            numCasillaCarcel = carcel;
        } else {
            numCasillaCarcel = 1;
        }
        casillas = new ArrayList<>();
        Casilla salida = new Casilla("Salida", TipoCasilla.DESCANSO);
        casillas.add(salida);
        porSalida = 0;
        tieneJuez = false;
    }
    
    private boolean correcto(){     //Comprueba la validez del tablero
        if (casillas.size() > numCasillaCarcel && tieneJuez) {
            return true;
        } else {
            return false;
        }
    }
    
    private boolean correcto(int numCasilla){       //Comprueba la validez de una casilla
        return correcto() && casillas.size() > numCasilla;
    }
    
    int getCarcel() {       //Consultor del numCasillaCarcel
        return numCasillaCarcel;
    }
    
    int getPorSalida() {    //Devuelve el valor de porSalida
        if (porSalida > 0 ) {
            int valor = porSalida;
            porSalida--;
            return valor;
        } else {
            return porSalida;
        }
    }
    
    void añadeCasilla(Casilla casilla) {        //Añade una casilla al tablero
        if (casillas.size() == numCasillaCarcel) {
            Casilla carcel = new Casilla("Cárcel",TipoCasilla.DESCANSO);
            casillas.add(carcel);
        }
        casillas.add(casilla);
        if (casillas.size() == numCasillaCarcel) {
            Casilla carcel = new Casilla("Cárcel", TipoCasilla.DESCANSO);
            casillas.add(carcel);
        }
    }
    
    void añadeJuez() {          //Añade el juez al tablero
        if (!tieneJuez) {
            Casilla juez = new Casilla("Juez", TipoCasilla.JUEZ);
            casillas.add(juez);
            tieneJuez = true;
        }
    }
    
    Casilla getCasilla(int numCasilla) {        //Devuelve la casilla solicitada
        if (correcto(numCasilla)) {
            return casillas.get(numCasilla);
        } else return null;
    }
    
    int nuevaPosicion(int actual, int tirada) {     //Calcula la nueva posicion del jugador en el tablero
        if (!correcto()) return -1;
        return (actual+tirada) % casillas.size();
    }
    
    int calcularTirada(int origen, int destino) {       //Calcula la tirada que tendrías que sacar para acabar en cierta posicion desde tu posicion actual
        int diferencia = destino-origen;
        if (diferencia < 0) {
            diferencia += casillas.size();
        }
        return diferencia;
    }
}

File: Segundo/PDOO/CivitasRuby/Civitas/Tablero.rb
# encoding:utf-8
require_relative "Casilla"

module Civitas

    class Tablero

        def initialize (c)  #Constructor con un parametro para saber cual será la casilla carcel
            @numCasillaCarcel
            if (c>=1)
                @numCasillaCarcel = c
            else
                @numCasillaCarcel = 1
            end
            @casillas = []
            casilla = Casilla.new(TipoCasilla::DESCANSO, "Salida")
            @casillas << casilla
            @porSalida = 0
            @tieneJuez = false
        end


        attr_reader :numCasillaCarcel
        

        def correcto    #Evalua si el tablero como conjunto es correcto para iniciar una partida
            return (@casillas.length > @numCasillaCarcel && @tieneJuez)
        end


        def correcta(numCasilla)    #Evalua si una casilla en concreto es correcta
            return (correcto() && numCasilla.to_i < @casillas.size)
        end


        def getPorSalida  #Devuelve el numero de veces que se ha pasado por la salida
            if(@porSalida > 0)
                @porSalida--
                valor = @porSalida + 1
                return valor
            else
                return @porSalida
            end
        end


        def añadeCasilla(c)     #Añade una casilla al tablero, comprobando antes que no sea en la posicion de la casilla carcel
            if (@casillas.size == @numCasillaCarcel)
                casilla = Civitas::Casilla.new(TipoCasilla::DESCANSO, "Cárcel")
                @casillas<< casilla
            end
            @casillas.push(c)
            if (@casillas.size == @numCasillaCarcel)
                casilla = Civitas::Casilla.new(TipoCasilla::DESCANSO, "Cárcel")
                @casillas << casilla
            end
        end

        def añadeJuez       #Añade el juez al tablero
            if(!@tieneJuez)
                juez = Casilla.new(TipoCasilla::JUEZ, "Juez")
                @casillas << juez
                @tieneJuez = true
            end
        end

        def getCasilla(numCasilla)    #Consultor de una casilla dada su posición en el tablero
            if (correcta(numCasilla))
                return @casillas.at(numCasilla)
            else
                return nil
            end
        end

        def nuevaPosicion(actual, tirada)   #Calcula la nueva posicion del jugador y la devuelve
            if (!correcto)
                return -1
            end
            return (actual+tirada) % @casillas.size
        end

        def calcularTirada(origen, destino)     #Calcula la tirada que habría sido necesaria para acabar en la posicion 'destino'
            resultado = destino-origen
            if (resultado < 0)
                resultado += @casillas.size
            end
            return resultado
        end
        
        
    end
end